### DISCUSSION TITLE
Please don't deprecate $use and $on('query')

### DISCUSSION THREAD

**Discussion by gamedevsam** [prisma/prisma#20234] [Q&A]
Some of the new initiatives with extensions are great, but I don't like to see these useful features get deprecated while the community and ecosystem around extensions is still quite immature.

I appreciate all the changes the Prisma is doing, but also believe the team should take a more considerate / delibrate approach when it comes to deprecating features widely used by the community.

Please re-consider deprecating `$use`.

**Comment by janpio** [comment]
Don't worry, [middlewares](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) are not going away for now.

We just think that [Client Extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions) are the better solution moving forward, as many of the problems with middlewares we actually can't solve there, but don't even have or can overcome with Client Extensions. For that reason, we feel it is only fair to indicate this via a deprecation of middlewares.

What "more deliberate" approach were you thinking of?

**Reply by gamedevsam** [reply]
> What "more deliberate" approach were you thinking of?

I don't know... deprecate, but keep it around until there is consensus among community to delete?

**Reply by janpio** [reply]
Consensus does not work in this case. We need to be able to drive Prisma ORM forward, which can mean removing parts we can not improve and maintain anymore. Consensus, or even democratic voting, robs us that ability as Prisma to direct where we are going with Prisma ORM.

We will not remove middlewares before all existing use cases are fully covered. If that takes a few major releases, that will be the earliest we remove middlewares. But I don't actually think it will take multiple major releases until Client Extensions are clearly better and can do everything middlewares can, and more. And then there is no reason, besides existing middlewares that have not been transformed into Client Extensions yet, to not make that step.

**Reply by gamedevsam** [reply]
Fair enough, I am happy with your response and would consider that a "deliberate" approach to the deprecations.

**Comment by masterbater** [comment]
> Some of the new initiatives with extensions are great, but I don't like to see these useful features get deprecated while the community and ecosystem around extensions is still quite immature.
> 
> I appreciate all the changes the Prisma is doing, but also believe the team should take a more considerate / delibrate approach when it comes to deprecating features widely used by the community.
> 
> Please re-consider deprecating `$use`.

I believe it should be removed immediately and not be deprecated. It has weird sequence behavior and middleware is limited. New client extensions are very intuitive and wont also prevent you to make different prisma clients with your modified. Very powerful, modular, it makes things way easy to customize depending on the scope you want per prisma client

**Reply by janpio** [reply]
We will never remove such important and widely used functionality without prior deprecation. There is no harm in keeping existing middlewares working and around, but the deprecation makes it clear that we suggest using Client Extensions for future approaches.

**Reply by masterbater** [reply]
> We will never remove such important and widely used functionality without prior deprecation. There is no harm in keeping existing middlewares working and around, but the deprecation makes it clear that we suggest using Client Extensions for future approaches.

Yes release cycle needs deprecation before removing it to new major version but the modularity of client extensions makes it better for sharing and making libraries where removing is better since dependencies for existing libs are prisma 4.

**Reply by gamedevsam** [reply]
My concern is that now that it's been deprecated it would be removed during next major release. And that makes me sad, because at least for me and other Nest users the current implementation of extensions (or maybe just its TypeScript support) has some issues (see [this issue for more details](https://github.com/prisma/prisma/issues/18628)).

**Reply by janpio** [reply]
We will not remove middlewares before Client Extensions are fully on par with what they can offer and support. Issues like the one you linked will be resolved first, before we even consider that. Although we are now on Prisma 5, since Prisma 2 we have only really broken existing functionality in a few minor cases. We won't change our approach now that we have deprecated middlewares.

**Comment by IbrahimMohammed47** [comment]
One place where extensions break but middleware don't is: https://github.com/olivierwilkinson/prisma-extension-soft-delete/issues/24#issuecomment-2217064730

That's why I want to keep having middleware in my project unless Prisma provides a native soft deletion feature

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
